server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java

```diff
@@ -70,59 +87,80 @@ public boolean acquireLock(List<RowLock> rowLocks) {
+                        filter(LambdaUtils.distinctByKey(LockDO::getRowKey))
+                        .collect(Collectors.toList());
             }
-            List<String> existedKeyList = new ArrayList<>();
-            locks.forEach(lockDO -> {
-                existedKeyList.add(getLockKey(lockDO.getRowKey()));
-            });
-            List<String> lockList = jedis.mget(existedKeyList.toArray(new String[0]));
-            Map<String, String> map = new HashMap<>(existedKeyList.size(), 1);
-            for (int i = 0; i < existedKeyList.size(); i++) {
-                String existedValue = lockList.get(i);
-                if (existedValue == null) {
-                    String key = existedKeyList.get(i);
-                    map.put(key, JSON.toJSONString(locks.get(i)));
+            List<String> needLockKeys = new ArrayList<>();
+            needLockDOS.forEach(lockDO -> needLockKeys.add(buildLockKey(lockDO.getRowKey())));
+
+            Pipeline pipeline1 = jedis.pipelined();
+            needLockKeys.stream().forEachOrdered(needLockKey -> pipeline1.hget(needLockKey, XID));
+            List<Object> existedXidObjs = pipeline1.syncAndReturnAll();
+            List<String> existedXids = (List<String>) (List) existedXidObjs;
+            Map<String, LockDO> map = new HashMap<>(needLockKeys.size(), 1);
+
+            String needLockXid = rowLocks.get(0).getXid();
+            for (int i = 0; i < needLockKeys.size(); i++) {
+                String existedXid = existedXids.get(i);
+                if (StringUtils.isEmpty(existedXid)) {
+                    //If empty,we need to lock this row
+                    map.put(needLockKeys.get(i), needLockDOS.get(i));
                 } else {
-                    LockDO existed = JSON.parseObject(existedValue, LockDO.class);
-                    if (!StringUtils.equals(existed.getXid(), locks.get(i).getXid())) {
+                    if (!StringUtils.equals(existedXid, needLockXid)) {
+                        //If not equals,means the rowkey is holding by another global transaction
                         return false;
                     }
                 }
             }
+
             if (map.isEmpty()) {
                 return true;
             }
             Pipeline pipeline = jedis.pipelined();
             List<String> readyKeys = new ArrayList<>();
             map.forEach((key, value) -> {
-                pipeline.setnx(key, value);
+                Map<String, String> lockMap = new HashMap<>(8);
+                lockMap.put(XID, value.getXid());
+                lockMap.put(TRANSACTION_ID, value.getTransactionId().toString());
+                lockMap.put(BRANCH_ID, value.getBranchId().toString());
+                lockMap.put(RESOURCE_ID, value.getResourceId());
+                lockMap.put(TABLE_NAME, value.getTableName());
+                lockMap.put(ROW_KEY, value.getRowKey());
+                lockMap.put(PK, value.getPk());
+                pipeline.hmset(key, lockMap);
```

Comment:
cannot play the role of competition lock

Reply:
fixed.